types/src/addressable_entity.rs

```diff
@@ -608,6 +609,113 @@ impl KeyValueJsonSchema for EntryPointLabels {
     const JSON_SCHEMA_KV_NAME: Option<&'static str> = Some("NamedEntryPoint");
 }
 
+/// Collection of named message topics.
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize, Debug, Default)]
+#[cfg_attr(feature = "datasize", derive(DataSize))]
+#[cfg_attr(feature = "json-schema", derive(JsonSchema))]
+#[serde(transparent, deny_unknown_fields)]
+pub struct MessageTopics(
+    #[serde(with = "BTreeMapToArray::<String, TopicNameHash, MessageTopicLabels>")]
+    BTreeMap<String, TopicNameHash>,
+);
+
+impl ToBytes for MessageTopics {
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        self.0.to_bytes()
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.0.serialized_length()
+    }
+
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), bytesrepr::Error> {
+        self.0.write_bytes(writer)
+    }
+}
+
+impl FromBytes for MessageTopics {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (message_topics_map, remainder) = BTreeMap::<String, TopicNameHash>::from_bytes(bytes)?;
+        Ok((MessageTopics(message_topics_map), remainder))
+    }
+}
+
+impl MessageTopics {
+    /// Adds new message topic by topic name.
+    pub fn add_topic(
+        &mut self,
+        topic_name: String,
+        topic_name_hash: TopicNameHash,
+    ) -> Result<(), MessageTopicError> {
+        if self.0.len() > u32::MAX as usize {
```

Comment:
Shouldn't this use `MessagesLimits::max_topics_per_contract` and `>=` instead of `>`?

Reply:
Indeed it should be `>=` but `MessagesLimits::max_topics_per_contract` should be enforced at a higher level (which it is). This is just a hard limit that is enforced.

Reply:
changed here: https://github.com/casper-network/casper-node/pull/4315/commits/3190647ec0012448907208e3a8406244433c5486